### PR TITLE
[W5.4][T09-A4] Niu Yunpeng

### PR DESCRIPTION
**Pull request by yunpengn**
- Add and fully support `update` command
- In OOP style
- Update tests and documentation
- Make sure all attributes of contact details are optional to `update` , for a better UX (user experience).

Yunpeng

**Updated: Ready-for-review**
- I have finished IO test updating.

### CODE REVIEW

File: src/seedu/addressbook/commands/AddCommand.java [hunk 1]
```diff
@@ -1,19 +1,13 @@
...
-
 import seedu.addressbook.data.exception.IllegalValueException;
-import seedu.addressbook.data.person.Address;
-import seedu.addressbook.data.person.Email;
-import seedu.addressbook.data.person.Name;
-import seedu.addressbook.data.person.Person;
-import seedu.addressbook.data.person.Phone;
-import seedu.addressbook.data.person.ReadOnlyPerson;
-import seedu.addressbook.data.person.UniquePersonList;
+import seedu.addressbook.data.person.*;
```
Comment (Houjisan) [inline src/seedu/addressbook/commands/AddCommand.java]: Is using the wildcard for importing recommended over explicitly listing each and every class to import?

File: src/seedu/addressbook/commands/UpdateCommand.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    private final UniqueTagList tagList;
+
+    public UpdateCommand(int targetVisibleIndex, String name,
+                         String phone, boolean isPhonePrivate,
+                         String email, boolean isEmailPrivate,
+                         String address, boolean isAddressPrivate,
+                         Set<String> tags) throws IllegalValueException {
+        super(targetVisibleIndex);
+
+        this.name = isBeingUpdate(name) ? new Name(name) : null;
```
Comment (Houjisan) [inline src/seedu/addressbook/commands/UpdateCommand.java]: isBeingUpdated might sound a bit better

File: src/seedu/addressbook/data/AddressBook.java [hunk 3]
```diff
@@ -53,7 +52,7 @@ public AddressBook(UniquePersonList persons, UniqueTagList tags) {
      *  - exists in the master list {@link #allTags}
      *  - points to a Tag object in the master list
      */
-    private void syncTagsWithMasterList(Person person) {
+    private void syncTagsWithMasterList(ReadOnlyPerson person) {
```
Comment (Houjisan) [inline src/seedu/addressbook/data/AddressBook.java]: Agree with the change to use interface instead

File: src/seedu/addressbook/commands/AddCommand.java (hunk 1 above)
Comment (yunpengn) [reply src/seedu/addressbook/commands/AddCommand.java]: Yes, agree with this. I also think that wildcard should be encouraged in Java import statement.

However, when I use Intellij IDEA built-in "import optimization" tool, it automatcially changes to this way.

Need further clarification.

File: src/seedu/addressbook/commands/UpdateCommand.java (hunk 2 above)
Comment (yunpengn) [reply src/seedu/addressbook/commands/UpdateCommand.java]: yes, already update the method name

File: src/seedu/addressbook/commands/AddCommand.java (hunk 1 above)
Comment (ngsikai) [reply src/seedu/addressbook/commands/AddCommand.java]: Even thought Intellij has a built-in functionality of collapsing multiple imports into wildcard for convenience and more concise files, it is still clearer to explicit list out dependencies within a class as it make tracing someone else's code a lot easier. 

Some preliminary searching on the web brought me to this article on how to configure against the "import optimization": https://stackoverflow.com/questions/3348816/intellij-never-use-wildcard-imports

File: src/seedu/addressbook/commands/FindCommand.java [hunk 4]
```diff
@@ -1,13 +1,9 @@
...
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Set;
-
 import seedu.addressbook.data.person.ReadOnlyPerson;
 
+import java.util.*;
+
```
Comment (ngsikai) [inline src/seedu/addressbook/commands/FindCommand.java]: Same point on import optimization and its impact on explicitness.

File: src/seedu/addressbook/commands/AddCommand.java (hunk 1 above)
Comment (yunpengn) [reply src/seedu/addressbook/commands/AddCommand.java]: Noted with thanks.

Will try to avoid "Intellij"izing the import statement.

**Pull request closed without merging**